File: 112-path-sum/112-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    let sum = 0;
    let boolean = false;
    
    (function DFS(root, count) {
        if (root === null || boolean === true) return;
        count += root.val
        if (root.left === null && root.right === null) {
            boolean = count === targetSum ? true : false 
        }
        DFS(root.left, count)
        DFS(root.right, count)
    })(root, 0);
    
    return boolean;    
    //declare sum variable = 0
    //declare boolean var is false
    
    //traverse tree using DFS
    //if encounter a leaf node
    //if sum is target sum
    //boolean is true
    
    //return boolean
};

File: 112-path-sum/112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        
        targetSum -= root.val;
        
        if ((root.left == null) && (root.right == null)) {
            return (targetSum == 0);
        }
        
        return hasPathSum(root.left, targetSum) || hasPathSum(root.right, targetSum);     
    }
}
/*
check if root is null, return if so
subtract val from targetSum
if root left and root right are both null
return if targetSum = 0
return recursively searching left or right subtrees returning true
*/
